libsolidity/interface/UniversalCallback.h

```diff
@@ -0,0 +1,53 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+#pragma once
+
+#include <libsolidity/interface/FileReader.h>
+#include <libsolidity/interface/SMTSolverCommand.h>
+
+namespace solidity::frontend
+{
+
+/// UniversalCallback is used to wrap both FileReader and SMTSolverCommand
+/// callbacks in a single callback.  It uses the Kind of the callback to
+/// determine which to call internally.
+class UniversalCallback
+{
+public:
+	UniversalCallback(FileReader&  _fileReader, SMTSolverCommand& _solver) :
+		m_fileReader{_fileReader},
+		m_solver{_solver}
+	{}
+
+	frontend::ReadCallback::Callback callback()
+	{
+		return [this](std::string const& _kind, std::string const& _data) -> ReadCallback::Result {
```

Comment:
Is there a reason `_kind` could not be `ReadCallbackKind::Kind` instead of a string?

Reply:
let me check

Reply:
good point, a few places are doing `readFile(ReadCallback::kindString(ReadCallback::Kind::ReadFile), ...)` for the `FileReader` callback which looks silly, so will change to `Kind` directly.

Reply:
Sure thing - however if this requires changing it in too many places, feel free to open a separate issue for this in case you don't want to pollute this PR with a refactor.

Reply:
I think the reason was to keep the callback generic actually, as in you can pass in any kind you want, cus ultimately this callback may call something outside the compiler (the callback can be given via the C interface of the compiler, which `solcjs` uses also for file stuff). So I wouldn't change it in the `ReadCallback` itself, but here it might make sense to change it.

Reply:
Generic is fine, but passing the callback type as a string is just weird to me.

Reply:
So, this requires deeper changes in `FileReader` and `ReadFile` as well, so I wouldn't touch all that in this PR if it's fine.

Reply:
Yup, sounds good, don't worry about.